plenoptic/metric/perceptual_distance.py

```diff
@@ -250,6 +245,83 @@ def ssim_map(img1, img2, dynamic_range=1):
+    an improvement upon structural similarity index (SSIM) that takes into
+    account the perceptual distance between two images on different scales.
+
+    SSIM is based on three comparison measurements between the two images:
+    luminance, contrast, and structure. All of these are computed in windows
+    across the images, producing three maps instead of scalars. The SSIM map is
+    the elementwise product of these three maps. See `metric.ssim` and
+    `metric.ssim_map` for a full description of SSIM.
+
+    To get images of different scales, average pooling operations with kernel
+    size 2 are performed recursively on the input images. The product of
+    contrast map and structure map (the "contrast-structure map") is computed
+    for all but the coarsest scales, and the overall SSIM map is only computed
+    for the coarsest scale. Their mean values are raised to exponents and
+    multiplied to produce MS-SSIM:
+    .. math::
+        MSSSIM = {SSIM}_M^{a_M} \prod_{i=1}^{M-1} ({CS}_i)^{a_i}
+    Here :math: `M` is the number of scales, :math: `{CS}_i` is the mean value
+    of the contrast-structure map for the i'th finest scale, and :math: `{SSIM}_M`
+    is the mean value of the SSIM map for the coarsest scale. If at least one
+    of these terms are negative, the value of MS-SSIM is zero.
+
+    Parameters
+    ----------
+    img1 : torch.Tensor
+        4d tensor with first image to compare
+    img2 : torch.Tensor
+        4d tensor with second image to compare. Must have the same height and
+        width (last two dimensions) as `img1`
+    dynamic_range : int, optional.
+        dynamic range of the images. Note we assume that both images have the
+        same dynamic range. 1, the default, is appropriate for float images
+        between 0 and 1, as is common in synthesis. 2 is appropriate for float
+        images between -1 and 1, and 255 is appropriate for standard 8-bit
+        integer images. We'll raise a warning if it looks like your value is
+        not appropriate for `img1` or `img2`, but will calculate it anyway.
+    power_factors : 1D array, optional.
+        power exponents for the mean values of maps, for different scales (from
+        fine to coarse). The length of this array determines the number of scales.
+        By default, this is set to [0.0448, 0.2856, 0.3001, 0.2363, 0.1333],
+        which is what psychophysical experiments in [1]_ found.
+
+    Returns
+    ------
+    msssim : torch.Tensor
+        2d tensor of shape (batch, channel) containing the MS-SSIM for each image
+
+    References
+    ----------
+    .. [1] Wang, Zhou, Eero P. Simoncelli, and Alan C. Bovik. "Multiscale
+       structural similarity for image quality assessment." The Thrity-Seventh
+       Asilomar Conference on Signals, Systems & Computers, 2003. Vol. 2. IEEE, 2003.
+
+    """
+    if power_factors is None:
+        power_factors = [0.0448, 0.2856, 0.3001, 0.2363, 0.1333]
```

Comment:
these particular values have been calibrated for images where 5 scales can be computed, so I think it would be nice to assert the size of the input images passed to this function and make sure they are of the expected overall size. Images of shape 512×512 (even 256x256) or bigger should be fine, but computing SSIM on the smallest scale for images smaller than that will not be possible

Reply:
The SSIM is supposed to use a 11x11 Gaussian convolutional window, but currently supports height and width smaller than 11 by setting the window size to `min(11, height, width)`. Through this, MS-SSIM also currently supports smaller images. I can add a warning for SSIM when height or width is smaller than 11, and a warning for MS-SSIM when the coarsest scale has height or width smaller than 11. Or I can consider these situations not worth supporting and raise an exception for these situations. What do you think?

Reply:
I support the adding of warning for small images for when e.g. log2(imsize)<5

Reply:
When images are really large, do these numbers still make sense? For example, if the image is 2048 x 2048, we only compute the components on the first 5 scales, right? And so effectively ignore the coarsest info?

Reply:
I think it still makes sense. The way I think about this is that if a 512x512 image and a 2048x2048 image are displayed with the same visual angle to a viewer, then the scales that should be computed are the same.

Reply:
sounds goodt o me. I'm going to mark this as resolved (since I think the warning you added covers @lyndond and @pehf 's comments, but they should re-open this if not)